plugins/inputs/openstack/openstack.go

```diff
@@ -0,0 +1,908 @@
+		if err := o.gatherHypervisors(acc); err != nil {
+			acc.AddError(fmt.Errorf("failed to get resource hypervisors %v", err))
+		}
+	}
+	server_gather := o.InEnabledServices("servers")
+	for _, hypervisor := range o.hypervisors {
+		page, err := servers.List(o.compute, &servers.ListOpts{AllTenants: true, Host: hypervisor.HypervisorHostname}).AllPages()
+		if err != nil {
+			return fmt.Errorf("unable to list servers %v", err)
+		}
+		servers, err := servers.ExtractServers(page)
+		if err != nil {
+			return fmt.Errorf("unable to extract servers %v", err)
+		}
+		for _, server := range servers {
+			if server_gather {
+				tags := map[string]string{}
+				// Extract the flavor details to avoid joins (ignore errors and leave as zero values)
+				var vcpus, ram, disk int
+				if flavorIDInterface, ok := server.Flavor["id"]; ok {
+					if flavorID, ok := flavorIDInterface.(string); ok {
+						tags["flavor"] = flavorID
+						if flavor, ok := o.flavors[flavorID]; ok {
+							vcpus = flavor.VCPUs
+							ram = flavor.RAM
+							disk = flavor.Disk
+						}
+					}
+				}
+				if imageIDInterface, ok := server.Image["id"]; ok {
+					if imageID, ok := imageIDInterface.(string); ok {
+						tags["image"] = imageID
+					}
+				}
+				// Try derive the associated project
+				project := "unknown"
+				if p, ok := o.projects[server.TenantID]; ok {
+					project = p.Name
+				}
+				tags["tenant_id"] = server.TenantID
+				tags["name"] = server.Name
+				tags["host_id"] = server.HostID
+				tags["status"] = strings.ToLower(server.Status)
+				tags["key_name"] = server.KeyName
+				tags["host_name"] = hypervisor.HypervisorHostname
+				tags["project"] = project
+				fields := map[string]interface{}{
+					"id":               server.ID,
+					"progress":         server.Progress,
+					"accessIPv4":       server.AccessIPv4,
+					"accessIPv6":       server.AccessIPv6,
+					"addresses":        len(server.Addresses),
+					"security_groups":  len(server.SecurityGroups),
+					"volumes_attached": len(server.AttachedVolumes),
+					"fault_code":       server.Fault.Code,
+					"fault_details":    server.Fault.Details,
+					"fault_message":    server.Fault.Message,
+					"fault_created":    server.Fault.Created.Format("2006-01-02T15:04:05.999999999Z07:00"),
+					"updated":          server.Updated.Format("2006-01-02T15:04:05.999999999Z07:00"),
+					"created":          server.Created.Format("2006-01-02T15:04:05.999999999Z07:00"),
+					"vcpus":            vcpus,
+					"ram_mb":           ram,
+					"disk_gb":          disk,
+				}
+				if o.OutputSecrets {
+					tags["user_id"] = server.UserID
+					fields["adminPass"] = server.AdminPass
+				}
+				if len(server.AttachedVolumes) == 0 {
+					acc.AddFields("openstack_server", fields, tags)
+				} else {
+					for _, AttachedVolume := range server.AttachedVolumes {
+						fields["volume_id"] = AttachedVolume.ID
+						acc.AddFields("openstack_server", fields, tags)
+					}
+				}
+			}
```

Comment:
How about moving the inner of the `for` loop to a separate function (e.g. `gatherServerDetails()`) and then do
```golang
		if server_gather {
			for _, server := range servers {
				o.gatherServerDetails(acc, server)
			}
		}
```